docs/how-to/h-upgrade.md

```diff
@@ -0,0 +1,115 @@
+1. Adding or removing units
+2. Creating or destroying new relations
+3. Changes in workload configuration
+   4. Upgrading other connected applications (e.g. ZooKeeper)
+
+The concurrency with other operations is not supported, and it can lead the cluster into inconsistent states.
+
+## Minor upgrade process overview
+
+When performing an in-place upgrade process, the full process is composed by the following high-level steps:
+
+1. **Collect** all necessary pre-upgrade information, necessary for a rollback (if ever needed)
+2. **Prepare** the charm for the in-place upgrade, by running some preparatory tasks 
+3. **Upgrade** the charm and/or the workload. Once started, all units in a cluster will refresh the charm code and undergo a workload restart/update. The upgrade will be aborted if the unit upgrade has failed, requiring the admin user to rollback.
+4. **Post-upgrade checks** to make sure all units are in the proper state and the cluster is healthy.
+
+## Step 1: Collect
+
+The first step is to record the revisions of the running application, as a safety measure for a rollback action if needed. To accomplish this, simply run the `juju status` command and look for the revisions of the deployed Kafka and Zookeeper applications. You can also retrieve this with the following command (that requires [`yq`](https://snapcraft.io/install/yq/ubuntu) to be installed):
+
+```shell
+KAFKA_CHARM_REVISION=$(juju status --format json | yq .applications.<KAFKA_APP_NAME>.charm-rev)
+ZOOKEEPER_CHARM_REVISION=$(juju status --format json | yq .applications.<ZOOKEEPER_APP_NAME>.charm-rev)
+```
+
+Please fill `<KAFKA_APP_NAME>` and `<ZOOKEEPER_APP_NAME>}` placeholder appropriately, e.g. `kafka` and `zookeeper`.
+
+## Step 2: Prepare
+
+Before upgrading, the charm needs to perform some preparatory tasks to define the upgrade plan.  
+To do so, run the `pre-upgrade-check` action against the leader unit:
+
+```shell
+juju run-action kafka/leader pre-upgrade-check --wait
+```
+
+Make sure that the output of the action is successful, e.g. the output should read:
+
+```shell
+unit-kafka-0:
+  ...
+  results: {}
+  status: completed
+  ...
+```
+
+> Note that you won't be able to upgrade successfully unless you complete successfully this action. 
+The action will also configure the charm to minimize high-availability reduction and ensure a safe upgrade process. After successful execution, the charm is ready to be upgraded.
+
+## Step 3: Upgrade
+
+Use the [`juju refresh`](https://juju.is/docs/juju/juju-refresh) command to trigger the charm upgrade process.
+Note that the upgrade can be performed against:
+
+* selected channel/track, therefore upgrading to the latest revision published on that track
+```shell
+juju refresh kafka --channel 3/edge
+```
+* selected revision
+```shell
+juju refresh kafka --revision=<REVISION>
```

Comment:
JFYI: https://bugs.launchpad.net/juju/+bug/2033048

Reply:
Thanks for the link! I will just leave the reference, as hopefully this bug is fixed soon (I see that they put importance as "High")